server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java

```diff
@@ -297,27 +303,20 @@ protected void produceMail(Map<String, Object> props, int msgPrio, Mail mail) th
         props.put(JAMES_MAIL_MESSAGE_SIZE, mail.getMessageSize());
         props.put(JAMES_MAIL_NAME, mail.getName());
 
-        StringBuilder recipientsBuilder = new StringBuilder();
-
-        Iterator<MailAddress> recipients = mail.getRecipients().iterator();
-        while (recipients.hasNext()) {
-            String recipient = recipients.next().toString();
-            recipientsBuilder.append(recipient.trim());
-            if (recipients.hasNext()) {
-                recipientsBuilder.append(JAMES_MAIL_SEPARATOR);
-            }
+        // won't serialize the empty headers so it is mandatory
+        // to handle nulls when reconstructing mail from message
+        if (!mail.getPerRecipientSpecificHeaders().getHeadersByRecipient().isEmpty()) {
+            byte[] serialize = SerializationUtils.serialize(mail.getPerRecipientSpecificHeaders());
+            props.put(JAMES_MAIL_PER_RECIPIENT_HEADERS, Hex.encodeHexString(serialize));
         }
-        props.put(JAMES_MAIL_RECIPIENTS, recipientsBuilder.toString());
+
+        String recipientsBuilder = Joiner.on(JAMES_MAIL_SEPARATOR).skipNulls().join(mail.getRecipients());
+
+        props.put(JAMES_MAIL_RECIPIENTS, recipientsBuilder);
         props.put(JAMES_MAIL_REMOTEADDR, mail.getRemoteAddr());
         props.put(JAMES_MAIL_REMOTEHOST, mail.getRemoteHost());
 
-        String sender;
-        MailAddress s = mail.getSender();
-        if (s == null) {
-            sender = "";
-        } else {
-            sender = mail.getSender().toString();
-        }
+        String sender = Optional.ofNullable(mail.getSender()).map(MailAddress::toString).orElse("");
```

Comment:
Use `asString` instead of `toString`.

toString generally serves for debugging, and it's value can change...

Reply:
Sure, I just used old code, it seem to me bit difficult to read. I spent some time understand 5 lines and fill bit confused.

Reply:
The new "Optional" version looks great and is way easier to read, yes, thank you.